test/functional/ex_cmds/dict_notifications_spec.lua

```diff
@@ -253,7 +253,20 @@ describe('dictionary change notifications', function()
       command('call g:ReplaceWatcher2()')
       command('let g:key = "value"')
       eq({'notification', '2b', {'key', {old = 'v2', new = 'value'}}}, next_msg())
+    end)
+  end)
 
+  describe('with lambdas', function()
+    it('works correctly', function()
+      source([[
+      let g:myDict = {'foo': 'baz'}
+      let dict = g:myDict
+      let key = 'foo'
+      let value = 'bar'
+      call dictwatcheradd(g:myDict, 'foo', dict, key, value -> { rpcnotify(g:channel, '2', key, value) })
+      let g:myDict.foo = 'bar'
+      ]])
```

Comment:
I think this could probably be simplified to:

```vim
let dict = {'foo': 'baz'}
call dictwatcheradd(dict, 'foo', {dict, key, value -> rpcnotify(g:channel, '2', key, value)})
let dict.foo = 'bar'
```